<html><head></head><body>
  <div id="_idContainer411">
    <h1 class="chapterNumber">17</h1>
    <h1 id="_idParaDest-334" class="chapterTitle">Genetic Algorithms in Hybrid Neural Networks</h1>
    <p class="normal">In this chapter and the following two chapters, we will explore the world inside us. First, in this chapter, we will use the model of our genes as an optimizing tool. In <em class="italics">Chapter 18</em>, <em class="italics">Neuromorphic Computing</em>, we will enter our biological brain activity and create neuromorphic networks. Finally, in <em class="italics">Chapter 19</em>, <em class="italics">Quantum Computing</em>, we will go even deeper and use the quantum material in us to build quantum mechanic models for quantum computing.</p>
    <p class="normal">A slight change in any of these tiny entities (genes, neurons, qubits) within us can modify our whole existence.</p>
    <p class="normal">In this chapter, we will discover how to enter our chromosome, find our genes, and understand how our reproduction process works. From there, we will begin to implement an evolutionary algorithm in Python, a <strong class="bold">genetic algorithm</strong> (<strong class="bold">GA</strong>).</p>
    <p class="normal">Charles Darwin offered "survival of the fittest" as a model to represent evolution. In some ways, the model is controversial. In 21st century societies, we tend to provide support to those who are not the fittest, as best as possible. However, in mathematics, we do not have this ethical problem.</p>
    <p class="normal">In AI, we need to provide an accurate solution. If we generate several solutions, we can apply the "survival of the fittest" to abstract numbers.</p>
    <p class="normal">In some cases, GAs dramatically reduce the number of combinations required to find the optimal solution to a problem. By generating mathematical offspring, choosing the fittest, and producing new stronger abstract generations, the system often reaches a more optimal solution than propagating permutations. </p>
    <p class="normal">A well-designed GA can optimize the architecture of a neural network, thereby producing a hybrid neural network.</p>
    <p class="normal">The following topics will be covered in this chapter:</p>
    <ul>
      <li class="list">Evolutionary algorithms; genetic algorithms</li>
      <li class="list">Extending the genes of genetic algorithms to optimizing tools</li>
      <li class="list">Hybrid neural networks</li>
      <li class="list">Using a genetic algorithm to optimize an LSTM</li>
    </ul>
    <p class="normal">Let's begin by first working to understand what evolutionary algorithms are.</p>
    <h1 id="_idParaDest-335" class="title">Understanding evolutionary algorithms</h1>
    <p class="normal">In this section, we will drill down<a id="_idIndexMarker821"/> from our heredity down to our genes to understand the process that we will then represent while building our Python program.</p>
    <p class="normal">Successive generations of humans activate some genes and not others, producing the wonderful diversity of humanity. A human lifetime is an episode in a long line of thousands of generations of humans. We all have two parents, four grandparents, and eight great-grandparents, which amounts to 2<sup>3</sup> ascendants. Suppose that we extend this line of reasoning to four generations per century and then over about 12,000 years when the last glacial period ended and the planet started warming up. We obtain:</p>
    <ul>
      <li class="list">4 * 1 century * 10 centuries = 1,000 years and 40 generations</li>
      <li class="list">40 generations * 12= 480</li>
      <li class="list">Adding up to 2<sup>480</sup> mathematical ascendants to anybody living today on the planet!</li>
    </ul>
    <p class="normal">Even if we limit ourselves to 1,000 years, 2<sup>40</sup>, that adds up to 1,099,511,627,776 ascendants a thousand years ago. But there is a problem. This figure is impossible! Today, we have reached the height of the human population, which is only 7,500,000,000. So, this means that our ascendants had many children who married their cousins of all degrees, making humanity one large extended family, no matter what our skin color or hair color is!</p>
    <h2 id="_idParaDest-336" class="title">Heredity in humans</h2>
    <p class="normal">First, men <a id="_idIndexMarker822"/>fertilize women when a male cell unites with a female cell. The fertilized egg grows and, after quite an adventure, is born and becomes one of us writing or reading this book.</p>
    <p class="normal">To grow and live from being a fertilized egg to human adults, we must gather much of the outside world, absorb, and transform it.</p>
    <p class="normal">This transformation of food and materials by us until we grow into something that is more or less like our ancestors is called heredity.</p>
    <p class="normal">If life were calm and quiet, nothing noticeable would occur. However, our environment has exerted relentless pressure on us for millions of years, back to when we were just some kind of bacteria floating around in an ocean somewhere. That pressure brought about continuous natural selection; what worked would live on, whilst what didn't work would die out.</p>
    <p class="normal">That pressure continues up to present day, forcing us, humans, to adapt genetically or disappear. Those humans who failed to adapt to face the pressure of their environment died out. We who live today have survived.</p>
    <p class="normal">Evolution can be <a id="_idIndexMarker823"/>defined as a state of constant conflict. On one hand we have our relentless, often hostile environment. On the other, our equally relentless genes; many of which die out, but others morph, adapt, and continue on through heredity—indeed, they're doing that even now, and who knows what is going to happen next?</p>
    <h3 id="_idParaDest-337" class="title">Our cells</h3>
    <p class="normal">In our cells, the nucleus<a id="_idIndexMarker824"/> contains personal biological data in the form of chromosomes. Our cells contain 46 chromosomes per cell, which, in turn, are formed by 23 pairs of chromosomes. One of these pairs is a sex cell to determine our sex.</p>
    <p class="normal">Inside the chromosomes, we have genes, especially the mitochondrial genome, which is the DNA in tiny cells that take our food and transform it into fuel for our cells. Each cell is a microscopic busy factory containing thousands of genes!</p>
    <p class="normal">The human genome describes an incredible set of sequences for our building blocks contained in the 23 chromosomes of our cells.</p>
    <h3 id="_idParaDest-338" class="title">How heredity works</h3>
    <p class="normal">Except for the <a id="_idIndexMarker825"/>sex cells, we inherit twenty-three of our mother's forty-six chromosomes and twenty-three of our father's chromosomes. In turn, our parents' cells contain the chromosomes of their parents—our grandparents—and so on in various proportions.</p>
    <p class="normal">Let's take an example to view the complexity of what we are facing. We will take one of our parents, either our mother or father. <em class="italics">P</em> represents the set of chromosomes of that parent. </p>
    <p class="normal">Letters with primes will represent their fathers'—our grandfathers'—chromosomes, and letters with double primes will represent their mothers'—our grandmothers'—chromosomes.</p>
    <p class="normal">We can represent this as follows for your father in his sex cell:</p>
    <figure class="mediaobject"><img src="../Images/B15438_17_001.png" alt=""/></figure>
    <p class="normal">For your mother, the last chromosome of the first set would be an <em class="italics">X</em> in her sex cell:</p>
    <figure class="mediaobject"><img src="../Images/B15438_17_002.png" alt=""/></figure>
    <p class="normal">Women are <em class="italics">X</em>-<em class="italics">X</em> and men <em class="italics">X</em>-<em class="italics">Y</em>.</p>
    <p class="normal">Imagine the possibilities!</p>
    <p class="normal">If we only take <em class="italics">A</em>, <em class="italics">B</em>, and <em class="italics">C</em> in only one of our parent's cells, we already obtain the following set, <em class="italics">C</em>, of the eight combinations we would inherit:</p>
    <figure class="mediaobject"><img src="../Images/B15438_17_003.png" alt=""/></figure>
    <p class="normal">If we extend this to <a id="_idIndexMarker826"/>the twenty-three chromosomes, the distribution climbs up to 2<sup>23</sup>, or 8,388,608 possibilities.</p>
    <p class="normal">Our evolutionary process contains the right potential for evolutionary algorithms.</p>
    <h2 id="_idParaDest-339" class="title">Evolutionary algorithms</h2>
    <p class="normal">In this section, we will drill<a id="_idIndexMarker827"/> down further into evolutionary algorithms, getting closer to our Python programs. Evolutionary algorithms can be used in any field in which combinations are useful: scheduling, medical research on DNA, weather forecasting, neural network architecture optimizing and a limitless number of domains.</p>
    <p class="normal"><em class="italics">Evolutionary computation</em> is <a id="_idIndexMarker828"/>a set of algorithms that apply trial-and-error techniques to reproduce an abstract mathematical version of biological evolution. This mathematical model does not contend with having solved the difficult task of explaining evolution, which naturally cannot be reduced to a few equations.</p>
    <p class="normal">However, our biological environment produces frameworks that, though the fruits of our imagination, enable us to create efficient abstract algorithms.</p>
    <p class="normal">Evolutionary algorithms <a id="_idIndexMarker829"/>enter the category of <em class="italics">evolutionary computation</em>. An evolutionary algorithm contains processes such as mutation, crossover, and selection. Many models can achieve the goals set for an evolutionary process.</p>
    <p class="normal">A GA introduces the category of an evolutionary algorithm.</p>
    <p class="normal">We will first define the concepts involved in a GA, which are the building blocks of a Python program.</p>
    <h3 id="_idParaDest-340" class="title">Going from a biological model to an algorithm</h3>
    <p class="normal">There are many ways of creating a <a id="_idIndexMarker830"/>GA model. You can reproduce the exact description of the human model described in the preceding section, you can simplify it, or you can create another view.</p>
    <p class="normal">Our model will contain a set of genes in a chromosome and a population to interact with:</p>
    <figure class="mediaobject"><img src="../Images/B15438_17_01.png" alt=""/></figure>
    <p class="packt_figref">Figure 17.1: Chromosomes in genes</p>
    <p class="normal">Our model is an abstract model for algorithms, not the actual representation of human chromosomes that come in pairs, for example. A gene in an individual's chromosome will interact with the gene set of the population. This process will be defined in the following sections.</p>
    <h3 id="_idParaDest-341" class="title">Basic concepts</h3>
    <p class="normal">Let's first describe the concepts of the building blocks of our Python program:</p>
    <ul>
      <li class="list"><strong class="bold">Target</strong>: Target <a id="_idIndexMarker831"/>defines the length and properties of the child we wish to obtain after <em class="italics">n</em> generations.<p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">In our model, the target can be specified or unspecified.</p>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">A specified target contains a length and its value:</p>
        <pre class="programlisting"><code class="hljs cmake">    <span class="hljs-keyword">target</span> = <span class="hljs-string">"Algorithm"</span> <span class="hljs-comment"># No space unless specified as a character in the gene set</span>
</code></pre>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">In this case, our environment requires an exact gene set to remain fit in that environment. See the following fitness function.</p>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">An unspecified target contains a length, but not its actual value, which must be found by the GA after <em class="italics">n</em> generations:</p>
        <pre class="programlisting"><code class="hljs cmake">    <span class="hljs-keyword">target</span>=<span class="hljs-string">"AAAA"</span> <span class="hljs-comment">#unspecified target</span>
</code></pre>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">In this case, our environment does not require an exact gene, but rather a gene with the features required to remain fit in that environment. See the following fitness function.</p>
      </li>
      <li class="list"><strong class="bold">Population</strong>: Population<a id="_idIndexMarker832"/> first defines the selection of an individual at random we will call <strong class="bold">parent</strong> that will contain a certain length of its string of genes:
        <pre class="programlisting"><code class="hljs ruby"><span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">gen_parent</span><span class="hljs-params">(length)</span></span>
</code></pre>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">Population also defines the potential individuals the parent can interact with to produce a child. In our genetic simulation, the size of the population is represented by the gene set of the population (see the following point).</p>
      </li>
      <li class="list"><strong class="bold">Gene set of the parent</strong>: The gene set<a id="_idIndexMarker833"/> of the parent will first be a random set of genes.</li>
      <li class="list"><strong class="bold">Gene set of the population</strong>: This <a id="_idIndexMarker834"/>parent will then randomly encounter another person in the population with random gene choices. This population is represented by a gene set that we will draw genes from randomly:
        <pre class="programlisting"><code class="hljs ini"><span class="hljs-attr">geneSet</span> = <span class="hljs-string">"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!.-"</span>
</code></pre>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">Since a given population can possess any or several of these genes, our population is very large from the start.</p>
      </li>
      <li class="list"><strong class="bold">Selection</strong>: We will <a id="_idIndexMarker835"/>select genes randomly from our gene set to simulate the meeting of our parent with another parent:
        <pre class="programlisting"><code class="hljs sas">    <span class="hljs-meta">index</span>=random.ran<span class="hljs-meta">drange(</span>0,l<span class="hljs-meta">en(</span>parent))
</code></pre>
      </li>
      <li class="list"><strong class="bold">Crossover</strong>: We will <a id="_idIndexMarker836"/>then form a child from the parent and the random gene taken from our population represented by the gene set.<p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">In our model, once the random selection of the gene has been made, the gene will be replaced by a random gene taken from the population represented by the gene set of the population:</p>
        <pre class="programlisting"><code class="hljs haxe">    <span class="hljs-keyword">if</span>(<span class="hljs-keyword">new</span><span class="hljs-type">Gene</span>!=oldGene):<span class="hljs-type">childGenes</span>[index]=<span class="hljs-keyword">new</span><span class="hljs-type">Gene</span>;
</code></pre>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">Notice that the new gene must be different from the old gene during the crossover in our algorithm. Our algorithm thus avoids getting stuck in some local combination during the reproduction phase. Child after child, we will produce new generations.</p>
      </li>
      <li class="list"><strong class="bold">Mutation</strong>: We <a id="_idIndexMarker837"/>will not accept the same gene from a parent to form a child. If we detect this, we will randomly change a gene to make sure each generation is different.<p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">As described in the <em class="italics">Crossover</em> paragraph, we will not accept a child with the same genes as a parent. As seen in the heredity section of this chapter, it is unlikely that a child would inherit exactly all of the genes of a given parent. <em class="italics">Diversity</em> is the key to producing generation after generation of children that will adapt to their environment.</p>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">In our model, a diversity rule forcing mutation has been introduced:</p>
        <pre class="programlisting"><code class="hljs haxe">    <span class="hljs-keyword">if</span>(<span class="hljs-keyword">new</span><span class="hljs-type">Gene</span>==oldGene):<span class="hljs-type">childGenes</span>[index]=alternate;
</code></pre>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">We thus introduce an alternate gene, as we will see while building the Python program.</p>
      </li>
      <li class="list"><strong class="bold">Child</strong>: Child defines<a id="_idIndexMarker838"/> a set of genes that contains the same number of genes as the parent but with new genes. Generations of children of child gene strings will be produced and then complete the selection process with a fitness function.</li>
      <li class="list"><strong class="bold">Fitness</strong>: Fitness<a id="_idIndexMarker839"/> defines the value of the child as defined in a given model. The fittest will then be selected to be the parent for the next generation.<p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">In our model, we have two fitness functions that we define in a scenario variable.</p>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">If <code class="Code-In-Text--PACKT-">scenario=1</code>, then a specified target scenario will be activated. The target will be specified to fit the surrounding environment.</p>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">Polar bears became white, for example, to blend in with the surrounding snow and ice of their environment. In other areas, bears are often brown, for example, to blend in with the surrounding vegetation.</p>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">The fitness function thus has a target to reach. For example, in one instance in the program:</p>
        <pre class="programlisting"><code class="hljs aspectj">    <span class="hljs-keyword">target</span>=<span class="hljs-string">"FBDC"</span>
</code></pre>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">This target could mean many things, as we will see. In the case of the bears, maybe these genes in one of their chromosomes trigger off their color: white or brown.</p>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">If our program simulating nature does not produce the right gene in a given generation, this means that the bear is not mutating correctly and will not survive. The fitness function of the Python program simulates nature by only keeping the strings of genes that make the child evolve in the right direction.</p>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">The specified target, in this case, is an identified string of genes that will make a life-and-death difference.</p>
        <pre class="programlisting"><code class="hljs ruby"><span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">get_fitness</span><span class="hljs-params">(this_choice,scenario)</span></span>:
    <span class="hljs-keyword">if</span>(scenario==<span class="hljs-number">1</span>):
</code></pre>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">As we will see <a id="_idIndexMarker840"/>later, <code class="Code-In-Text--PACKT-">scenario==1</code> will make sure that the exact gene set required to survive is reached.</p>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">If <code class="Code-In-Text--PACKT-">scenario=0</code>, then an unspecified target will be activated. The length of the target is not specified. However, a feature set will define the value of the gene set of the population. This feature contains numerical values that open the door to any optimization you wish to solve, as we will see in the Python program example. This numerical set is called the KPI set because the values are key performance indicators of the system we will explore in this model:</p>
        <pre class="programlisting"><code class="hljs ini"><span class="hljs-attr">KPIset</span> =<span class="hljs-string">"0123456772012345674701234569980923456767012345671001234"</span>   <span class="hljs-comment">#KPI set</span>
</code></pre>
        <p class="Bullet-Without-Bullet-Within-Bullet-End--PACKT-">The <code class="Code-In-Text--PACKT-">KPIset</code> feature set matches the size of the gene set of the population.</p>
      </li>
    </ul>
    <h2 id="_idParaDest-342" class="title">Building a genetic algorithm in Python</h2>
    <p class="normal">We will now build a <a id="_idIndexMarker841"/>GA from scratch using <code class="Code-In-Text--PACKT-">GA.ipynb</code>.</p>
    <p class="normal">You can use the <code class="Code-In-Text--PACKT-">.py</code> version. The code is the same, although the line numbers in this chapter refer to the Jupyter notebook cells of the <code class="Code-In-Text--PACKT-">.ipynb</code> version.</p>
    <div class="note">
      <p class="Information-Box--PACKT-">At all times, you can go back to the previous section, <em class="italics">Basic concepts</em>, to consult the definitions used to describe the Python program in the following sections.</p>
    </div>
    <h3 id="_idParaDest-343" class="title">Importing the libraries</h3>
    <p class="normal">This program is built <a id="_idIndexMarker842"/>from scratch with no higher-level library to get the feel of a GA. Three lines are enough to get everything working:</p>
    <pre class="programlisting"><code class="hljs elm"><span class="hljs-keyword">import</span> math
<span class="hljs-keyword">import</span> random
<span class="hljs-keyword">import</span> datetime
</code></pre>
    <h3 id="_idParaDest-344" class="title">Calling the algorithm</h3>
    <p class="normal">In this program, we<a id="_idIndexMarker843"/> will be exploring three scenarios. Two scenarios generate specified targets, and one generates an unspecified target.</p>
    <p class="normal">We will begin with a specified target and then move to the more advanced unspecified target that will prepare us for hybrid networks using a GA.</p>
    <p class="normal">We will first go to the <em class="italics">Calling the Algorithm</em> cell of the program. The first task is to define which scenario and type of fitness function we will use on line 3:</p>
    <pre class="programlisting"><code class="hljs routeros"><span class="hljs-attribute">scenario</span>=0   # <span class="hljs-attribute">1</span>=target provided at start, <span class="hljs-attribute">0</span>=<span class="hljs-literal">no</span> target, genetic optimizer
</code></pre>
    <p class="normal">If <code class="Code-In-Text--PACKT-">scenario=1</code>, the program generates the correct exact genes for a child from several generations, starting with a random seed parent.</p>
    <p class="normal">If <code class="Code-In-Text--PACKT-">scenario=0</code>, the program generates the best features of a type of genes of a child from several generations, starting with a random seed parent.</p>
    <p class="normal">In line 4, <code class="Code-In-Text--PACKT-">GA=2</code> defines which target we are dealing with.</p>
    <p class="normal">If <code class="Code-In-Text--PACKT-">GA=1</code>, the gene set of the population and the target are defined. The main function is called:</p>
    <pre class="programlisting"><code class="hljs nix"><span class="hljs-keyword">if</span>(<span class="hljs-attr">GA==1):</span>
    <span class="hljs-attr">geneSet</span> =
        <span class="hljs-string">"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!.-"</span>
    <span class="hljs-comment"># target with no space unless specified as a character in the geneSet</span>
    <span class="hljs-attr">target</span> = <span class="hljs-string">"Algorithm"</span>  <span class="hljs-comment"># No space unless specified as a character in the geneSet</span>
    print(<span class="hljs-string">"geneSet:"</span>,geneSet,<span class="hljs-string">"\n"</span>,<span class="hljs-string">"target:"</span>,target)
    ga_main()
</code></pre>
    <p class="normal">At this point, the gene set of the population is printed along with the target:</p>
    <pre class="programlisting"><code class="hljs dts"><span class="hljs-symbol">geneSet:</span> abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!.-
<span class="hljs-symbol"> target:</span> Algorithm
</code></pre>
    <p class="normal">The last line calls <code class="Code-In-Text--PACKT-">ga_main()</code>, the main function.</p>
    <h3 id="_idParaDest-345" class="title">The main function</h3>
    <p class="normal">The <a id="_idIndexMarker844"/>code is in the <code class="Code-In-Text--PACKT-">ga_main()</code> cell of the program.</p>
    <p class="normal">The <code class="Code-In-Text--PACKT-">ga_main()</code> function is divided into three parts: parent generation, child generation, and a summary.</p>
    <h3 id="_idParaDest-346" class="title">The parent generation process</h3>
    <p class="normal">The<a id="_idIndexMarker845"/> parent generation runs from line 2 to line 7:</p>
    <pre class="programlisting"><code class="hljs routeros">    <span class="hljs-attribute">startTime</span>=datetime.datetime.now()
    <span class="hljs-builtin-name">print</span>(<span class="hljs-string">"starttime"</span>,startTime)
    <span class="hljs-attribute">alphaParent</span>=gen_parent(len(target))
    <span class="hljs-attribute">bestFitness</span>=get_fitness(alphaParent,scenario)
    display(alphaParent,bestFitness,bestFitness,startTime) #<span class="hljs-literal">no</span> childFitness <span class="hljs-keyword">for</span> generation 1 so<span class="hljs-built_in"> default </span>value sent
</code></pre>
    <ul>
      <li class="list"><code class="Code-In-Text--PACKT-">startTime</code> indicates the start time, which is printed:
        <pre class="programlisting"><code class="hljs angelscript">starttime <span class="hljs-number">2019</span><span class="hljs-number">-10</span><span class="hljs-number">-12</span> <span class="hljs-number">10</span>:<span class="hljs-number">32</span>:<span class="hljs-number">28.294943</span>
</code></pre>
      </li>
      <li class="list"><code class="Code-In-Text--PACKT-">alphaParent</code> is the first parent who will be created by the <code class="Code-In-Text--PACKT-">gen_parent</code> function, which will be described in the following Jupyter notebook cell.</li>
      <li class="list"><code class="Code-In-Text--PACKT-">bestFitness</code> is the fitness estimated by the <code class="Code-In-Text--PACKT-">get_fitness</code> function, which will be described in this chapter.</li>
      <li class="list"><code class="Code-In-Text--PACKT-">display</code> is the function that describes the result of this process.</li>
    </ul>
    <p class="normal">We now have a process that creates a parent: <code class="Code-In-Text--PACKT-">gen_parent</code>, <code class="Code-In-Text--PACKT-">get_fitness</code>, and <code class="Code-In-Text--PACKT-">display</code>.</p>
    <p class="normal">We will now explore the parent generation function before resuming the main function.</p>
    <h3 id="_idParaDest-347" class="title">Generating a parent</h3>
    <p class="normal">The parent generation <a id="_idIndexMarker846"/>function cell starts with zero genes and the length of the target:</p>
    <pre class="programlisting"><code class="hljs ruby"><span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">gen_parent</span><span class="hljs-params">(length)</span></span>:
</code></pre>
    <p class="normal"><code class="Code-In-Text--PACKT-">length</code> = the length of the target. The target is <code class="Code-In-Text--PACKT-">"Algorithm"</code>, so <code class="Code-In-Text--PACKT-">length=9</code>.</p>
    <p class="normal">At the beginning of the process, the parent has no genes since the goal of this function is to produce a parent at random that contains a string of genes that is equal to the length of the target.</p>
    <p class="normal">The string of genes to produce a parent is as follows:</p>
    <pre class="programlisting"><code class="hljs nix">    <span class="hljs-attr">genes=[]</span>                        <span class="hljs-comment">#genes array</span>
</code></pre>
    <p class="normal">Now, a <code class="Code-In-Text--PACKT-">while</code> loop starts on line 3 of the cell to fill <code class="Code-In-Text--PACKT-">genes[]</code> until it reaches the target's length:</p>
    <pre class="programlisting"><code class="hljs livecodeserver">    <span class="hljs-keyword">while</span> <span class="hljs-built_in">len</span>(genes)&lt;<span class="hljs-built_in">length</span>:  <span class="hljs-comment">#genes is constrained to the length</span>
        <span class="hljs-comment">#sampleSize: length of target constraint</span>
        sampleSize=<span class="hljs-built_in">min</span>(<span class="hljs-built_in">length</span>-<span class="hljs-built_in">len</span>(genes),<span class="hljs-built_in">len</span>(geneSet))
        <span class="hljs-comment">#extend genes with a random sample the size of sampleSize extracted from geneSet</span>
        genes.extend(<span class="hljs-built_in">random</span>.sample(geneSet,sampleSize))
</code></pre>
    <ul>
      <li class="list"><code class="Code-In-Text--PACKT-">sampleSize</code> is the sample of genes required from the gene set, <code class="Code-In-Text--PACKT-">geneSet</code>, from which to choose a random gene for the parent.</li>
      <li class="list"><code class="Code-In-Text--PACKT-">genes.extend</code> adds a random gene to the <code class="Code-In-Text--PACKT-">genes</code> array from <code class="Code-In-Text--PACKT-">geneSet</code>.</li>
    </ul>
    <p class="normal">Once the parent's <a id="_idIndexMarker847"/>gene set, <code class="Code-In-Text--PACKT-">genes[]</code>, has reached the target's length, the <code class="Code-In-Text--PACKT-">return</code> function sends the parent back to the main function, <code class="Code-In-Text--PACKT-">ga_main()</code>, where it is displayed with the <code class="Code-In-Text--PACKT-">display</code> function. The output of the parent in this random run was:</p>
    <pre class="programlisting"><code class="hljs ebnf"><span class="hljs-attribute">aFJPKzYBD</span>
</code></pre>
    <p class="normal">Naturally, the parent will be different at each run since this is a random process.</p>
    <p class="normal">The string of genes is now returned to the <code class="Code-In-Text--PACKT-">ga_main()</code> function:</p>
    <pre class="programlisting"><code class="hljs cs">    <span class="hljs-keyword">return</span> <span class="hljs-string">''</span>.<span class="hljs-keyword">join</span>(genes)
</code></pre>
    <p class="normal">Now, let's explore the fitness function and the <code class="Code-In-Text--PACKT-">display</code> function.</p>
    <h3 id="_idParaDest-348" class="title">Fitness</h3>
    <p class="normal">At this<a id="_idIndexMarker848"/> point in the <code class="Code-In-Text--PACKT-">ga_main()</code> function, the start time was printed out, and the parent created:</p>
    <pre class="programlisting"><code class="hljs routeros"><span class="hljs-comment">#I PARENT GENERATION</span>
    <span class="hljs-attribute">startTime</span>=datetime.datetime.now()
    <span class="hljs-builtin-name">print</span>(<span class="hljs-string">"starttime"</span>,startTime)
    <span class="hljs-attribute">alphaParent</span>=gen_parent(len(target))
</code></pre>
    <p class="normal">We need to evaluate the fitness of the parent before creating generations of children:</p>
    <pre class="programlisting"><code class="hljs reasonml">    bestFitness = get<span class="hljs-constructor">_fitness(<span class="hljs-params">alphaParent</span>,<span class="hljs-params">scenario</span>)</span>
</code></pre>
    <p class="normal">In this paragraph, we will only describe the specified target case, which is part of <code class="Code-In-Text--PACKT-">scenario=1</code>. We will create a fitness function with the target of a given choice. In this case, only the sum of correct genes is calculated:</p>
    <pre class="programlisting"><code class="hljs ceylon">def get<span class="hljs-number">_f</span>itness(<span class="hljs-keyword">this</span><span class="hljs-number">_</span>choice,scenario):
    <span class="hljs-keyword">if</span>(scenario==<span class="hljs-number">1</span>):
        fitness = sum(<span class="hljs-number">1</span> <span class="hljs-keyword">for</span> expected,
            <span class="hljs-keyword">actual</span> <span class="hljs-keyword">in</span> zip(target,<span class="hljs-keyword">this</span><span class="hljs-number">_</span>choice) <span class="hljs-keyword">if</span> expected==<span class="hljs-keyword">actual</span>)
</code></pre>
    <ul>
      <li class="list"><code class="Code-In-Text--PACKT-">this_choice</code> is the parent string of genes produced by the <code class="Code-In-Text--PACKT-">gen_parent</code> function described in the preceding code snippet.</li>
      <li class="list"><code class="Code-In-Text--PACKT-">scenario</code> indicates whether the function calculates the sum of fit genes or evaluates the features of the genes. In this case, the sum of the correct genes is calculated.</li>
      <li class="list">Fitness is the number of correct genes found when comparing the target, the expected value, to actual, the <code class="Code-In-Text--PACKT-">this_choice</code> variable.</li>
      <li class="list">If <code class="Code-In-Text--PACKT-">expected==actual</code>, the sum is incremented.</li>
      <li class="list"><code class="Code-In-Text--PACKT-">zip</code>, in Python, is an efficient feature that iterates over two lists at the same time.</li>
    </ul>
    <p class="normal">Once fitness, <a id="_idIndexMarker849"/>the sum of fit genes, is calculated, the function returns the value to the <code class="Code-In-Text--PACKT-">ga_main()</code> function:</p>
    <pre class="programlisting"><code class="hljs kotlin">    <span class="hljs-keyword">return</span> fitness
</code></pre>
    <p class="normal">The parent generation will now be displayed by a function called in <code class="Code-In-Text--PACKT-">main_ga()</code>.</p>
    <h3 id="_idParaDest-349" class="title">Display parent</h3>
    <p class="normal">At this point, <code class="Code-In-Text--PACKT-">ga_main()</code> has <a id="_idIndexMarker850"/>printed the start time, created a parent, and evaluated its fitness:</p>
    <pre class="programlisting"><code class="hljs routeros">def ga_main():
    #I PARENT GENERATION
    <span class="hljs-attribute">startTime</span>=datetime.datetime.now()
    <span class="hljs-builtin-name">print</span>(<span class="hljs-string">"starttime"</span>,startTime)
    <span class="hljs-attribute">alphaParent</span>=gen_parent(len(target))
    <span class="hljs-attribute">bestFitness</span>=get_fitness(alphaParent,scenario)
</code></pre>
    <p class="normal">The program will now display the basic information concerning the first generation: the parent generation called the <code class="Code-In-Text--PACKT-">display</code> function from <code class="Code-In-Text--PACKT-">main_ga()</code>, line 7:</p>
    <pre class="programlisting"><code class="hljs lisp">    display(<span class="hljs-name">alphaParent</span>,bestFitness,bestFitness,startTime)
</code></pre>
    <ul>
      <li class="list"><code class="Code-In-Text--PACKT-">alphaParent</code> is the gene string of the parent</li>
      <li class="list"><code class="Code-In-Text--PACKT-">bestFitness</code> is its fitness</li>
      <li class="list">Since there is no child yet, <code class="Code-In-Text--PACKT-">bestFitness</code> is sent as the default value of a child's fitness</li>
      <li class="list"><code class="Code-In-Text--PACKT-">startTime</code></li>
    </ul>
    <p class="normal">In the display parent cell, line 2, the <code class="Code-In-Text--PACKT-">display</code> function receives the data sent by <code class="Code-In-Text--PACKT-">main_ga()</code>:</p>
    <pre class="programlisting"><code class="hljs ruby"><span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">display</span><span class="hljs-params">(selection,bestFitness,childFitness,startTime)</span></span>:
</code></pre>
    <p class="normal">The <code class="Code-In-Text--PACKT-">display</code> function calculates the time it has taken and prints the information in a few lines:</p>
    <pre class="programlisting"><code class="hljs routeros">    <span class="hljs-attribute">timeDiff</span>=datetime.datetime.now()-startTime
    <span class="hljs-builtin-name">print</span>(<span class="hljs-string">"Selection:"</span>,selection,<span class="hljs-string">"Fittest:"</span>,bestFitness,
          <span class="hljs-string">"This generation Fitness:"</span>,childFitness,
          <span class="hljs-string">"Time Difference:"</span>,timeDiff)
</code></pre>
    <ul>
      <li class="list"><code class="Code-In-Text--PACKT-">selection</code> is the string of genes of this generation.</li>
      <li class="list"><code class="Code-In-Text--PACKT-">bestFitness</code> is the value of the best string of genes created up to now.</li>
      <li class="list"><code class="Code-In-Text--PACKT-">childFitness</code> is this generation's fitness value. The first generation is the value of the parent who has, for the moment, the fittest genes. The parent is the child of another parent, although this parent is the first generation we are taking into consideration.</li>
      <li class="list"><code class="Code-In-Text--PACKT-">timeDiff</code> is an important value when dealing with larger gene sets. It will help detect whether the algorithm is running well or reaching its limit.</li>
    </ul>
    <p class="normal">The output will be<a id="_idIndexMarker851"/> displayed for the parent generation and each generation that gets closer to the fittest generation defined by the target:</p>
    <pre class="programlisting"><code class="hljs angelscript">Selection: BnVYkFcRK Fittest: <span class="hljs-number">0</span> This generation Fitness: <span class="hljs-number">0</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.000198</span>
</code></pre>
    <p class="normal">This output will vary during each run of the program since this is a stochastic algorithm simulating the random events that occur in our natural and artificial environments.</p>
    <p class="normal">Before exploring the loop that creates an unlimited number of generations, let's build the crossover function.</p>
    <h3 id="_idParaDest-350" class="title">Crossover and mutation</h3>
    <p class="normal">Our model<a id="_idIndexMarker852"/> contains a <code class="Code-In-Text--PACKT-">crossover</code> function with a mutation rule to ensure diversity.</p>
    <p class="normal">The <code class="Code-In-Text--PACKT-">crossover</code> function starts with the parent.</p>
    <pre class="programlisting"><code class="hljs ruby"><span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">crossover</span><span class="hljs-params">(parent)</span></span>:
</code></pre>
    <p class="normal">Each child of each generation will become the parent of another child.</p>
    <p class="normal">As in nature, a random gene in a parent's gene will be selected to be replaced:</p>
    <pre class="programlisting"><code class="hljs livecodeserver">    index=<span class="hljs-built_in">random</span>.randrange(<span class="hljs-number">0</span>,<span class="hljs-built_in">len</span>(parent))<span class="hljs-comment">#producing a random position of the parent gene</span>
</code></pre>
    <p class="normal">The <code class="Code-In-Text--PACKT-">index</code> designates the exact location of the gene that will be replaced:</p>
    <figure class="mediaobject"><img src="../Images/B15438_17_01.png" alt=""/></figure>
    <p class="packt_figref">Figure 17.2: Chromosomes in genes</p>
    <p class="normal">We can see that gene <strong class="bold">D</strong> in the chromosome will be replaced by the gene <strong class="bold">z</strong> of an individual of the population.</p>
    <p class="normal">Now, we simulate the reproduction phase. The child inherits the genes of its parent:</p>
    <pre class="programlisting"><code class="hljs isbl">    <span class="hljs-variable">childGenes</span>=<span class="hljs-function"><span class="hljs-title">list</span>(<span class="hljs-variable">parent</span>)</span>
</code></pre>
    <p class="normal">The parent's string of genes is converted into a list with the <code class="Code-In-Text--PACKT-">list</code> function.</p>
    <p class="normal">The algorithm stores the parent's gene to replace in a variable:</p>
    <pre class="programlisting"><code class="hljs axapta">    oldGene=childGenes[<span class="hljs-keyword">index</span>]        <span class="hljs-meta"># for diversity check</span>
</code></pre>
    <p class="normal"><code class="Code-In-Text--PACKT-">oldGene</code> will be <a id="_idIndexMarker853"/>compared to the new gene generated to make sure that diversity is respected so as to avoid getting stuck in a local loop.</p>
    <p class="normal">A new gene is chosen at random in the gene set of the population to simulate the interaction of the child with a given person among an indefinite number of persons:</p>
    <pre class="programlisting"><code class="hljs haxe">    <span class="hljs-keyword">new</span><span class="hljs-type">Gene</span>,alternate=random.sample(geneSet,<span class="hljs-number">2</span>)
</code></pre>
    <p class="normal">Notice that at the same time, the new gene, <code class="Code-In-Text--PACKT-">newGene</code>, is selected at random, while an alternate gene, <code class="Code-In-Text--PACKT-">alternate</code>, is chosen as well. <code class="Code-In-Text--PACKT-">alternate</code> is chosen to replace <code class="Code-In-Text--PACKT-">newGene</code> to avoid making the wrong choice.</p>
    <p class="normal">If the new gene, <code class="Code-In-Text--PACKT-">newGene</code>, is not equal to the old gene, <code class="Code-In-Text--PACKT-">oldGene</code>, that the child can inherit:</p>
    <pre class="programlisting"><code class="hljs haxe">    <span class="hljs-keyword">if</span>(<span class="hljs-keyword">new</span><span class="hljs-type">Gene</span>!=oldGene):<span class="hljs-type">childGenes</span>[index]=<span class="hljs-keyword">new</span><span class="hljs-type">Gene</span>;  <span class="hljs-meta">#natural crossover</span>
</code></pre>
    <p class="normal">The new gene becomes part of the string of genes of the child.</p>
    <p class="normal">However, if <code class="Code-In-Text--PACKT-">newGene</code> is the same as <code class="Code-In-Text--PACKT-">oldGene</code>, this could compromise the whole genetic process, with generations of children that do not evolve. Also, the algorithm might get stuck or waste quite some time making the right selections.</p>
    <p class="normal">This is where the <a id="_idIndexMarker854"/>alternate gene comes in and becomes part of the string of genes of the child. This crossover rule and this alternate rule simulates the mutation process of this model.</p>
    <pre class="programlisting"><code class="hljs routeros">    <span class="hljs-keyword">if</span>(<span class="hljs-attribute">newGene</span>==oldGene):childGenes[index]=alternate;  #mutation introduced <span class="hljs-keyword">to</span> ensure diversity <span class="hljs-keyword">to</span> avoid <span class="hljs-keyword">to</span> <span class="hljs-builtin-name">get</span> stuck <span class="hljs-keyword">in</span> a local minima
</code></pre>
    <p class="normal">Diversity has been verified!</p>
    <p class="normal">The function now returns the new string of genes of the child so that its fitness value can be calculated:</p>
    <pre class="programlisting"><code class="hljs cs">    <span class="hljs-keyword">return</span> <span class="hljs-string">''</span>.<span class="hljs-keyword">join</span>(childGenes)
</code></pre>
    <h3 id="_idParaDest-351" class="title">Producing generations of children</h3>
    <p class="normal">At this point, we <a id="_idIndexMarker855"/>have generated the parent and explored the basic concepts and functions.</p>
    <p class="normal">We are ready for the generation loop. First, we will view the code, then represent it in a flowchart, and then describe the lines of code.</p>
    <p class="normal">Once the parent has been created, we enter the loop that simulates the evolutionary process for many generations:</p>
    <pre class="programlisting"><code class="hljs routeros">    <span class="hljs-keyword">while</span> <span class="hljs-literal">True</span>:
        g+=1
        <span class="hljs-attribute">child</span>=crossover(bestParent)        #mutation
        <span class="hljs-attribute">childFitness</span>=get_fitness(child,scenario) #number of correct genes
        <span class="hljs-keyword">if</span> bestFitness&gt;=childFitness:#
            continue
        display(child,bestFitness,childFitness,startTime)
        <span class="hljs-attribute">bestFitness</span>=childFitness
        <span class="hljs-attribute">bestParent</span>=child
        <span class="hljs-keyword">if</span> <span class="hljs-attribute">scenario</span>==1: <span class="hljs-attribute">goal</span>=len(alphaParent);#number of good <span class="hljs-attribute">genes</span>=parent length
        <span class="hljs-keyword">if</span> <span class="hljs-attribute">scenario</span>==0: <span class="hljs-attribute">goal</span>=threshold;
        <span class="hljs-keyword">if</span> childFitness&gt;=goal:
            break
</code></pre>
    <p class="normal">The loop is best represented in a flowchart:</p>
    <figure class="mediaobject"><img src="../Images/B15438_17_03.png" alt=""/></figure>
    <p class="packt_figref">Figure 17.3: Genetic algorithm flow chart</p>
    <p class="normal">The process of the <a id="_idIndexMarker856"/>flowchart is based on Python's <code class="Code-In-Text--PACKT-">continue</code> method:</p>
    <ol>
      <li class="list">The code that precedes the condition:<ul>
          <li class="Bullet-Within-Bullet--PACKT-">Increments the generation counter:</li>
        </ul>
        <pre class="programlisting"><code class="hljs angelscript">        g+=<span class="hljs-number">1</span>
</code></pre>
        <ul>
          <li class="Bullet-Within-Bullet--PACKT-">Calls the <code class="Code-In-Text--PACKT-">crossover</code> function to produce a child:</li>
        </ul>
        <pre class="programlisting"><code class="hljs isbl">        <span class="hljs-variable">child</span>=<span class="hljs-function"><span class="hljs-title">crossover</span>(<span class="hljs-variable">bestParent</span>)</span>
</code></pre>
        <ul>
          <li class="Bullet-Within-Bullet--PACKT-">Calls the fitness function to obtain the fitness value:</li>
        </ul>
        <pre class="programlisting"><code class="hljs reasonml">        childFitness=get<span class="hljs-constructor">_fitness(<span class="hljs-params">child</span>,<span class="hljs-params">scenario</span>)</span> #number <span class="hljs-keyword">of</span> correct genes
</code></pre>
      </li>
      <li class="list">The condition to see if the <code class="Code-In-Text--PACKT-">childFitness</code> is higher than the <code class="Code-In-Text--PACKT-">bestFitness</code> obtained:
        <pre class="programlisting"><code class="hljs armasm">        <span class="hljs-meta">if</span> <span class="hljs-keyword">bestFitness&gt;=childFitness:
</span></code></pre>
        <ul>
          <li class="Bullet-Within-Bullet--PACKT-">If the condition is <code class="Code-In-Text--PACKT-">True</code>, then the evolution must continue until a child is fitter than its parent. This sends the process back to the top of the <code class="Code-In-Text--PACKT-">while</code> loop.</li>
          <li class="Bullet-Within-Bullet--PACKT-">If the condition is <code class="Code-In-Text--PACKT-">False</code>, this means that the child is fitter than the parent, and then the code will go beyond the condition.</li>
        </ul>
      </li>
      <li class="list">The code beyond the condition and <code class="Code-In-Text--PACKT-">continue</code> method:<ul>
          <li class="Bullet-Within-Bullet--PACKT-">The code displays the child, the <code class="Code-In-Text--PACKT-">bestFitness</code> becomes the <code class="Code-In-Text--PACKT-">childFitness</code>, and the <code class="Code-In-Text--PACKT-">bestParent</code> is now the child:</li>
        </ul>
        <pre class="programlisting"><code class="hljs routeros">        display(child,bestFitness,childFitness,startTime)
        <span class="hljs-attribute">bestFitness</span>=childFitness
        <span class="hljs-attribute">bestParent</span>=child
</code></pre>
        <ul>
          <li class="Bullet-Within-Bullet--PACKT-">The goal of the two scenarios of our model is defined. The goal of <code class="Code-In-Text--PACKT-">scenario==1</code> is to reach the length of the target with the right genes. The goal of <code class="Code-In-Text--PACKT-">scenario==0</code> will be to reach a threshold we will define in the next section:</li>
        </ul>
        <pre class="programlisting"><code class="hljs routeros">        <span class="hljs-keyword">if</span> <span class="hljs-attribute">scenario</span>==1: <span class="hljs-attribute">goal</span>=len(alphaParent);
        <span class="hljs-keyword">if</span> <span class="hljs-attribute">scenario</span>==0: <span class="hljs-attribute">goal</span>=threshold;
</code></pre>
      </li>
      <li class="list">The <code class="Code-In-Text--PACKT-">break</code> condition of the loop:<p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">The evolutionary process will stop when the fittest child has been created after several generations containing the genes that meet the target:</p>
        <pre class="programlisting"><code class="hljs groovy">        <span class="hljs-keyword">if</span> childFitness&gt;=<span class="hljs-string">goal:</span>
            <span class="hljs-keyword">break</span>
</code></pre>
      </li>
    </ol>
    <p class="normal">The output of the <em class="italics">n</em> generations will <a id="_idIndexMarker857"/>appear as follows:</p>
    <pre class="programlisting"><code class="hljs angelscript">Genetic Algorithm
geneSet: abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!.-
 target: Algorithm
starttime <span class="hljs-number">2019</span><span class="hljs-number">-10</span><span class="hljs-number">-12</span> <span class="hljs-number">20</span>:<span class="hljs-number">47</span>:<span class="hljs-number">03.232931</span>
Selection: Xe!bMSRzV Fittest: <span class="hljs-number">0</span> This generation Fitness: <span class="hljs-number">0</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.000953</span>
Selection: Xe!bMSRhV Fittest: <span class="hljs-number">0</span> This generation Fitness: <span class="hljs-number">1</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.002404</span>
Selection: Xl!bMSRhV Fittest: <span class="hljs-number">1</span> This generation Fitness: <span class="hljs-number">2</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.004391</span>
Selection: XlgbMSRhV Fittest: <span class="hljs-number">2</span> This generation Fitness: <span class="hljs-number">3</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.006860</span>
Selection: XlgoMSRhV Fittest: <span class="hljs-number">3</span> This generation Fitness: <span class="hljs-number">4</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.009525</span>
Selection: AlgoMSRhV Fittest: <span class="hljs-number">4</span> This generation Fitness: <span class="hljs-number">5</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.011954</span>
Selection: AlgorSRhV Fittest: <span class="hljs-number">5</span> This generation Fitness: <span class="hljs-number">6</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.013745</span>
Selection: AlgorSthV Fittest: <span class="hljs-number">6</span> This generation Fitness: <span class="hljs-number">7</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.016339</span>
Selection: AlgorithV Fittest: <span class="hljs-number">7</span> This generation Fitness: <span class="hljs-number">8</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.019031</span>
Selection: Algorithm Fittest: <span class="hljs-number">8</span> This generation Fitness: <span class="hljs-number">9</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.022239</span>
</code></pre>
    <p class="normal">We can see the <a id="_idIndexMarker858"/>display of all of the generations as described in the <code class="Code-In-Text--PACKT-">display</code> function in the <em class="italics">Display parent</em> section of this chapter.</p>
    <h3 id="_idParaDest-352" class="title">Summary code</h3>
    <p class="normal">Once the evolutionary<a id="_idIndexMarker859"/> process is over, a summary code takes over:</p>
    <pre class="programlisting"><code class="hljs routeros"><span class="hljs-comment">#III. SUMMARY</span>
    <span class="hljs-builtin-name">print</span>(<span class="hljs-string">"Summary---------------------------------------------------"</span>)
    <span class="hljs-attribute">endTime</span>=datetime.datetime.now()
    <span class="hljs-builtin-name">print</span>(<span class="hljs-string">"endtime"</span>,endTime)
    <span class="hljs-builtin-name">print</span>(<span class="hljs-string">"geneSet:"</span>,geneSet);<span class="hljs-builtin-name">print</span>(<span class="hljs-string">"target:"</span>,target)
    <span class="hljs-builtin-name">print</span>(<span class="hljs-string">"geneSet length:"</span>,len(geneSet))
    <span class="hljs-builtin-name">print</span>(<span class="hljs-string">"target length:"</span>,len(target))
    <span class="hljs-builtin-name">print</span>(<span class="hljs-string">"generations:"</span>,g)
    <span class="hljs-builtin-name">print</span>(<span class="hljs-string">"Note: the process is stochastic so the number of generations will vary"</span>)
</code></pre>
    <p class="normal">The output for the example displayed for the evolutionary loop is:</p>
    <pre class="programlisting"><code class="hljs yaml"><span class="hljs-string">Summary---------------------------------------------------</span>
<span class="hljs-string">endtime</span> <span class="hljs-number">2019</span><span class="hljs-number">-10</span><span class="hljs-number">-12</span> <span class="hljs-number">20</span><span class="hljs-string">:47:03.257112</span>
<span class="hljs-attr">geneSet:</span> <span class="hljs-string">abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!.-</span>
<span class="hljs-attr">target:</span> <span class="hljs-string">Algorithm</span>
<span class="hljs-attr">geneSet length:</span> <span class="hljs-number">55</span>
<span class="hljs-attr">target length:</span> <span class="hljs-number">9</span>
<span class="hljs-attr">generations:</span> <span class="hljs-number">782</span>
<span class="hljs-attr">Note:</span> <span class="hljs-string">the</span> <span class="hljs-string">process</span> <span class="hljs-string">is</span> <span class="hljs-string">stochastic</span> <span class="hljs-string">so</span> <span class="hljs-string">the</span> <span class="hljs-string">number</span> <span class="hljs-string">of</span> <span class="hljs-string">generations</span> <span class="hljs-string">will</span> <span class="hljs-string">vary</span>
</code></pre>
    <div class="note">
      <p class="Information-Box--PACKT-">Before we move on, it is important to note that using permutations for 9 out of 55 elements, it would take 10 ** 15.36 calculations to reach the goal instead of the 782 generations in this example. GAs are thus a productive way of generating permutations.</p>
    </div>
    <p class="normal">We have now explored the core of the evolutionary process and the Python code. We will now build the unspecified target and optimizing code that will lead us to hybrid neural networks.</p>
    <h2 id="_idParaDest-353" class="title">Unspecified target to optimize the architecture of a neural network with a genetic algorithm</h2>
    <p class="normal">In this section, we are laying the grounds and motivation to optimize the architecture of neural networks through a hybrid neural network. The architecture of a physical neural network will be optimized by a GA.</p>
    <p class="normal">We will study a physical neural network and then see how to optimize its architecture with our GA.</p>
    <h3 id="_idParaDest-354" class="title">A physical neural network</h3>
    <p class="normal">We will begin<a id="_idIndexMarker860"/> with a physical network named S-FNN, a <strong class="bold">feedforward neural network </strong>(<strong class="bold">FNN</strong>). Please look very closely at the following figure and take as much time as necessary to understand its architecture:</p>
    <figure class="mediaobject"><img src="../Images/B15438_17_04.png" alt=""/></figure>
    <p class="packt_figref">Figure 17.4: Architecture of a neural network</p>
    <p class="normal">This physical network has some<a id="_idIndexMarker861"/> important specific features:</p>
    <ul>
      <li class="list">This is an FNN</li>
      <li class="list">There are three layers.</li>
      <li class="list">The neurons in each layer are not fully connected to the neurons of the next layer. The connections between the neurons of the three layers are not based on all of the possibilities. The connections between the layers are based upon statistics of the best connections in the past when running this network.<p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">The connections are the result of many runs of this network to determine which ones are the best.</p>
      </li>
      <li class="list">The connections <a id="_idIndexMarker862"/>represented are the most productive ones, leading to an output of <code class="Code-In-Text--PACKT-">1</code> and not <code class="Code-In-Text--PACKT-">0</code>. An output of <code class="Code-In-Text--PACKT-">1</code> is a success, while an output of <code class="Code-In-Text--PACKT-">0</code> is a failure.</li>
      <li class="list">The input is a random value of the dataset that must be classified as <code class="Code-In-Text--PACKT-">1</code> if it is a success and <code class="Code-In-Text--PACKT-">0</code> if it is a failure.</li>
      <li class="list">After a careful study of past runs of this network, it has been found that the productivity of the network fully relies, in this example, on the number and quality of the neurons in layer 2.</li>
    </ul>
    <p class="normal">Let's use what we have learned and apply it to the architecture of a soccer team.</p>
    <h3 id="_idParaDest-355" class="title">What is the nature of this mysterious S-FNN?</h3>
    <p class="normal">S-FNN<a id="_idIndexMarker863"/> is a soccer team's architecture! It is the representation of a 3-4-3 disposition of a soccer team before a given game.</p>
    <p class="normal">Like abstract neural networks, the number and quality of neurons we choose per layer are critical. In our abstract world of <strong class="bold">artificial neural networks </strong>(<strong class="bold">ANNs</strong>), we have many problems to solve. How many should we keep? How many should we eliminate through pooling layers? How many should we abandon through dropout layers? How do we really know? How long will trial and error take with large datasets and networks?</p>
    <p class="normal">Now, let's go back to the figure, read the following explanation, and start finding a method to solve the architectural complexity of a neural network:</p>
    <ul>
      <li class="list"><strong class="bold">Input</strong> represents the goal of best possible choices to input the ball in the game based on past game statistics.</li>
      <li class="list"><strong class="bold">Layer 1</strong> is the defense layer of three players represented by three neurons.</li>
      <li class="list"><strong class="bold">Layer 2</strong> is the middle field, the result of the transformation of layer 1's activity, the statistics of past games showing the fittest, the connections. The initial ball of the input has now been flowing through these layers game after game.</li>
      <li class="list"><strong class="bold">Layer 3</strong> is the attack layer, the one that will lead to the output classification; 1 for a success meaning a goal has been scored, or 0 for failure. The connections between layer 2 and layer 3 show the fittest results found game after game.</li>
    </ul>
    <p class="normal"><strong class="bold">The problem to solve</strong>: Layer 2, in our<a id="_idIndexMarker864"/> example, has been identified as the critical layer in this team. For years now, it has been a weak point. Let's see how our GA will help find the fittest players for this layer.</p>
    <p class="normal">We could run <code class="Code-In-Text--PACKT-">scenario==1</code> (specified target) with target <code class="Code-In-Text--PACKT-">ga==3</code> as follows:</p>
    <pre class="programlisting"><code class="hljs routeros"><span class="hljs-keyword">if</span>(<span class="hljs-attribute">scenario</span>==1 <span class="hljs-keyword">and</span> <span class="hljs-attribute">GA</span>==3):
    <span class="hljs-attribute">target</span>=<span class="hljs-string">"FBDC"</span>  # <span class="hljs-literal">No</span> space unless specified as a character <span class="hljs-keyword">in</span> the geneSet
    <span class="hljs-builtin-name">print</span>(<span class="hljs-string">"geneSet:"</span>,geneSet,<span class="hljs-string">"\n"</span>,<span class="hljs-string">"target:"</span>,target)
    ga_main()
</code></pre>
    <p class="normal">The <code class="Code-In-Text--PACKT-">geneSet</code> is the population of available players on the soccer market and <code class="Code-In-Text--PACKT-">target</code> is the string of genes we need for layer 2 of our physical network:</p>
    <pre class="programlisting"><code class="hljs properties"><span class="hljs-attr">Genetic</span> <span class="hljs-string">Algorithm</span>
<span class="hljs-attr">geneSet</span>: <span class="hljs-string">abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!.-</span>
<span class="hljs-attr">target</span>: <span class="hljs-string">FBDC</span>
</code></pre>
    <p class="normal">The target is found after 851 generations in this run.</p>
    <p class="normal">However, we would like the GA to find the players by itself based on the features with an unspecified target. Let's explore this step by step, cell by cell.</p>
    <h3 id="_idParaDest-356" class="title">Calling the algorithm cell</h3>
    <p class="normal">From lines 7 to 12, we defined the <a id="_idIndexMarker865"/>parameters of the target architecture we want for layer 2 of our network:</p>
    <ul>
      <li class="list"><code class="Code-In-Text--PACKT-">geneSet</code> is the set of all the available players for layer 2 of our network, whether they are in the team or on the market:
        <pre class="programlisting"><code class="hljs ini"><span class="hljs-attr">geneSet</span>=<span class="hljs-string">"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!.-"</span>   <span class="hljs-comment">#gene set</span>
</code></pre>
      </li>
      <li class="list"><code class="Code-In-Text--PACKT-">KPIset</code>, or the key performance indicator set, is a performance score per player on the market for middle field. Each integer of the set contains a score between 0 and 9 based on the statistics of a player when playing as a midfielder:
        <pre class="programlisting"><code class="hljs ini"><span class="hljs-attr">KPIset</span> =<span class="hljs-string">"0123456772012345674701234569980923456767012345671001234"</span>   <span class="hljs-comment">#KPI set</span>
</code></pre>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">The KPI set matches each member of the gene set.</p>
      </li>
      <li class="list"><code class="Code-In-Text--PACKT-">threshold</code> is the sum of the midfielders' performance we need to attain in order to have a strong midfield in the team.
        <pre class="programlisting"><code class="hljs ini"><span class="hljs-attr">threshold</span>=<span class="hljs-number">35</span>
</code></pre>
      </li>
    </ul>
    <p class="normal">Our goal is to find a string of genes based on their features, their KPI properties.</p>
    <p class="normal">The goal is to reach the threshold.</p>
    <p class="normal">We call our evolutionary function on line 23:</p>
    <pre class="programlisting"><code class="hljs routeros"><span class="hljs-keyword">if</span>(<span class="hljs-attribute">scenario</span>==0 <span class="hljs-keyword">and</span> <span class="hljs-attribute">GA</span>==2):
    <span class="hljs-attribute">target</span>=<span class="hljs-string">"AAAA"</span>                         #unspecified target
    <span class="hljs-builtin-name">print</span>(<span class="hljs-string">"geneSet:"</span>,geneSet,<span class="hljs-string">"\n"</span>,<span class="hljs-string">"target:"</span>,target)
    ga_main()
</code></pre>
    <p class="normal">As you can see, <code class="Code-In-Text--PACKT-">target</code> is set to a<a id="_idIndexMarker866"/> meaningless value that will rapidly evolve. The GA will have to find a fit sequence.</p>
    <p class="normal">We can skip the intermediate fitness functions we have described in the previous section and focus on the scenarios within the fitness cell.</p>
    <h3 id="_idParaDest-357" class="title">Fitness cell</h3>
    <p class="normal">We have already described the fitness function of the fitness cell in the previous sections. We will focus on the code of the unspecified target, <code class="Code-In-Text--PACKT-">scenario==0</code>.</p>
    <p class="normal">The first part of the scenario calculates the sum of the performance of each gene (potential midfielder):</p>
    <pre class="programlisting"><code class="hljs routeros">    <span class="hljs-keyword">if</span>(<span class="hljs-attribute">scenario</span>==0):
        <span class="hljs-attribute">cc</span>=list(this_choice) # cc= this choice
        <span class="hljs-attribute">gs</span>=list(geneSet)     # gene <span class="hljs-builtin-name">set</span>
        <span class="hljs-attribute">cv</span>=list(KPIset)      # value of each KPI <span class="hljs-keyword">in</span> the <span class="hljs-builtin-name">set</span>
        <span class="hljs-attribute">fitness</span>=0
        <span class="hljs-keyword">for</span> op1 <span class="hljs-keyword">in</span> range(0,len(geneSet)): #2.first <span class="hljs-builtin-name">find</span> parent gene <span class="hljs-keyword">in</span> gene <span class="hljs-builtin-name">set</span>
            <span class="hljs-keyword">for</span> op <span class="hljs-keyword">in</span> range(0,len(target)):
                <span class="hljs-keyword">if</span> cc[op]==gs[op1]:       #3.gene identified <span class="hljs-keyword">in</span> gene <span class="hljs-builtin-name">set</span>
                    <span class="hljs-attribute">vc</span>=int(cv[op1])       #4.value of critical path constraint
                    fitness+=vc
</code></pre>
    <p class="normal">The collective fitness of the sequence of genes (midfielders) is contained in the <code class="Code-In-Text--PACKT-">fitness</code> variable.</p>
    <p class="normal">However, we cannot accept the same gene twice, which would mean we have a clone of the midfielder on the field! So, we add some safety code to set <code class="Code-In-Text--PACKT-">fitness</code> to <code class="Code-In-Text--PACKT-">0</code> in that case:</p>
    <pre class="programlisting"><code class="hljs vim">            <span class="hljs-keyword">for</span> op in <span class="hljs-built_in">range</span>(<span class="hljs-number">0</span>,<span class="hljs-built_in">len</span>(target)):
                <span class="hljs-keyword">for</span> op1 in <span class="hljs-built_in">range</span>(<span class="hljs-number">0</span>,<span class="hljs-built_in">len</span>(target)):
                    <span class="hljs-keyword">if</span> op!=op1 <span class="hljs-built_in">and</span> <span class="hljs-keyword">cc</span>[op]==<span class="hljs-keyword">cc</span>[op1]:
                        fitness=<span class="hljs-number">0</span>    # <span class="hljs-keyword">no</span> repetitions allowed, mutation enforcement
</code></pre>
    <p class="normal">Now, we can go back <a id="_idIndexMarker867"/>to <code class="Code-In-Text--PACKT-">ga_main()</code> and complete our process.</p>
    <h3 id="_idParaDest-358" class="title">ga_main() cell</h3>
    <p class="normal">We have already described the fitness cell in the previous sections. Now, we will focus on the code of the unspecified target, <code class="Code-In-Text--PACKT-">scenario==0</code>.</p>
    <p class="normal">In the <code class="Code-In-Text--PACKT-">ga_main()</code> cell, we<a id="_idIndexMarker868"/> simply need to examine lines 22 to 24:</p>
    <pre class="programlisting"><code class="hljs routeros">        <span class="hljs-keyword">if</span> <span class="hljs-attribute">scenario</span>==0: <span class="hljs-attribute">goal</span>=threshold;
        <span class="hljs-keyword">if</span> childFitness&gt;=goal:
            break
</code></pre>
    <p class="normal">If the <code class="Code-In-Text--PACKT-">scenario==0</code>, <code class="Code-In-Text--PACKT-">childFitness</code> must be <code class="Code-In-Text--PACKT-">&gt;=goal</code> (sum of KPIs).</p>
    <p class="normal">We have found our midfielders!</p>
    <p class="normal">We will now display the result:</p>
    <pre class="programlisting"><code class="hljs angelscript">Genetic Algorithm
geneSet: abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!.-
 target: AAAA
starttime <span class="hljs-number">2019</span><span class="hljs-number">-10</span><span class="hljs-number">-13</span> <span class="hljs-number">09</span>:<span class="hljs-number">08</span>:<span class="hljs-number">21.043754</span>
Selection: PNVx Fittest: <span class="hljs-number">18</span> This generation Fitness: <span class="hljs-number">18</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.001252</span>
Selection: LNVx Fittest: <span class="hljs-number">18</span> This generation Fitness: <span class="hljs-number">24</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.004130</span>
Selection: LNVq Fittest: <span class="hljs-number">24</span> This generation Fitness: <span class="hljs-number">27</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.004708</span>
Selection: LNFq Fittest: <span class="hljs-number">27</span> This generation Fitness: <span class="hljs-number">29</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.010362</span>
Selection: LBFq Fittest: <span class="hljs-number">29</span> This generation Fitness: <span class="hljs-number">31</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.015797</span>
Selection: CBFq Fittest: <span class="hljs-number">31</span> This generation Fitness: <span class="hljs-number">33</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.024712</span>
Selection: CBFt Fittest: <span class="hljs-number">33</span> This generation Fitness: <span class="hljs-number">34</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.032680</span>
Selection: CBFD Fittest: <span class="hljs-number">34</span> This generation Fitness: <span class="hljs-number">35</span> Time Difference: <span class="hljs-number">0</span>:<span class="hljs-number">00</span>:<span class="hljs-number">00.048524</span>
Summary---------------------------------------------------
endtime <span class="hljs-number">2019</span><span class="hljs-number">-10</span><span class="hljs-number">-13</span> <span class="hljs-number">09</span>:<span class="hljs-number">08</span>:<span class="hljs-number">21.094005</span>
geneSet: abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!.-
target: AAAA
geneSet length: <span class="hljs-number">55</span>
target length: <span class="hljs-number">4</span>
generations: <span class="hljs-number">154</span>
Note: the process <span class="hljs-keyword">is</span> stochastic so the number of generations will vary
</code></pre>
    <p class="normal">In this case, the <a id="_idIndexMarker869"/>GA chose midfielders C, B, F, and D.</p>
    <p class="normal">We now have all of the ingredients, concepts, and knowledge required to understand an artificial hybrid neural network.</p>
    <h1 id="_idParaDest-359" class="title">Artificial hybrid neural networks</h1>
    <p class="normal">In the previous section, we <a id="_idIndexMarker870"/>used a GA to optimize a physical neural network.</p>
    <p class="normal">In this section, we will extend the concept of <em class="italics">hybrid</em> we have just explored to ANNs. The principle is the same, so it will be relatively easy, with the concepts you now have in mind, to intuitively grasp the RNN we will optimize in this section.</p>
    <p class="normal">The future of AI in society lies in the collective intelligence of humans (diversity), machines (AI and IoT), and nature (sustainable projects for our survival).</p>
    <p class="normal">In AI, this diversity lies in ensemble algorithms, meta-algorithms and hybrid systems. Deep learning has proven its point. We can create a neural network with TensorFlow 2.x in a few lines. However, more often than not, it takes days, weeks, and often months to fine-tune ANN models that rely on large amounts of data to provide a reliable model. And that's where hybrid neural networks are necessary.</p>
    <p class="normal">A deep learning network can use any form of another type of algorithm to become a more efficient system. In our case, we have chosen evolutionary algorithms that could be used in deep learning:</p>
    <ul>
      <li class="list">To improve the input by extracting sequences of data that fit the patterns we are looking for using features of the data to find the fittest data points</li>
      <li class="list">To optimize the weights of the layers of a network to boost the speed and power of traditional optimizers</li>
      <li class="list">To improve the classification phase of the output of a network by selecting the fittest solutions processed as a sequence of genes</li>
      <li class="list">To improve the tedious task of defining the architecture of a network</li>
    </ul>
    <p class="normal">In this section, we will focus on an example of optimizing the architecture of a network with a GA by:</p>
    <ul>
      <li class="list">Creating the RNN</li>
      <li class="list">Introducing a GA that will optimize the architecture of the network</li>
    </ul>
    <div class="note">
      <p class="Information-Box--PACKT-"><strong class="bold">Note</strong>: For information on genetic algorithm representations, please go back to the previous sections at all times to refresh the process of an evolutionary algorithm. Please also refer back to the previous chapters to consult the structure of a neural network, in particular, <em class="italics">Chapter 16</em>, <em class="italics">Improving the Emotional Intelligence Deficiencies of Chatbots</em>, which describes an RNN.</p>
    </div>
    <h2 id="_idParaDest-360" class="title">Building the LSTM</h2>
    <p class="normal">Open <code class="Code-In-Text--PACKT-">Genetic_Algorithm_LSTM.ipynb</code> in Google Colaboratory or Jupyter on your machine or <code class="Code-In-Text--PACKT-">genetic_algorithm_lstm.py</code>.</p>
    <p class="normal">The model of this <a id="_idIndexMarker871"/>education example is an LSTM.</p>
    <p class="normal">The goal will be to use the window size generated by the GA and run the LSTM with this window size. <strong class="bold">Root-mean-square error </strong>(<strong class="bold">RMSE</strong>) will be used to measure the fitness of the configuration.</p>
    <p class="normal">In short, RMSE first calculates the square of the difference between the training data and the testing data, like many similar formulas. Then, the root of the result is calculated. Just keep in mind that the RMSE will compare what we expect to get with what we actually get and produce a value.</p>
    <p class="normal">Let's now explore the main cells of the program:</p>
    <ul>
      <li class="list"><strong class="bold">Importing required packages cell</strong>: <code class="Code-In-Text--PACKT-">Genetic_Algorithm_LSTM.ipynb</code> starts by installing DEAP, an evolutionary computation framework:
        <pre class="programlisting"><code class="hljs cmake">!pip <span class="hljs-keyword">install</span> deap
</code></pre>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">We built a GA from scratch in the previous sections. We won't need to begin from nothing this time, since this program uses a framework.</p>
        <p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">Then, the program installs <code class="Code-In-Text--PACKT-">bitstring</code>, which helps to process binary data.</p>
      </li>
      <li class="list"><strong class="bold">Loading the data cell</strong>: The data wind power forecast data in <code class="Code-In-Text--PACKT-">train.csv</code> comes from <a href="https://www.kaggle.com/c/GEF2012-wind-forecasting/data"><span class="url">https://www.kaggle.com/c/GEF2012-wind-forecasting/data</span></a>.<p class="Bullet-Without-Bullet-Within-Bullet--PACKT-">The <code class="Code-In-Text--PACKT-">wp1</code> to <code class="Code-In-Text--PACKT-">wp7</code> columns provide normalized data collected from measurements of the wind power of seven wind farms. The goal of the LSTM will be to take the sequence of data and make wind power forecasts.</p>
      </li>
      <li class="list"><strong class="bold">Defining the basic functions cell</strong>: This cell prepares the dataset and trains the model in a standard process. We will focus on line 14:
        <pre class="programlisting"><code class="hljs angelscript">    window_size = window_size_bits.<span class="hljs-built_in">uint</span>
</code></pre>
      </li>
      <li class="list"><strong class="bold">Evolutionary model cell</strong>: The model uses the DEAP framework function, but we easily recognize the concepts we explored in the previous sections and that are<a id="_idIndexMarker872"/> initialized as follows:
        <pre class="programlisting"><code class="hljs ini"><span class="hljs-attr">population_size</span> = <span class="hljs-number">4</span>
<span class="hljs-attr">num_generations</span> = <span class="hljs-number">2</span>
<span class="hljs-attr">gene_length</span> = <span class="hljs-number">10</span>
</code></pre>
      </li>
    </ul>
    <p class="normal">The code is all set with a ready-to-use GA to optimize the window size for our network.</p>
    <h2 id="_idParaDest-361" class="title">The goal of the model</h2>
    <p class="normal">As in our previous section, the<a id="_idIndexMarker873"/> goal is to find the best window size for this network, just as we were looking for the best layer 2 in an earlier section.</p>
    <p class="normal">The model has done the following tasks:</p>
    <ul>
      <li class="list">Installed the packages, loaded the data, and found the window size of the LSTM</li>
      <li class="list">Then, it ran the GA model to test the possible window size of the LSTM</li>
      <li class="list">An RMSE measurement is provided with each generation and production of epochs</li>
    </ul>
    <p class="normal">The program takes some time to run, but the results I ran are saved in the notebook so that you can view them. Here is one result to view how the system works:</p>
    <pre class="programlisting"><code class="hljs angelscript">Epoch <span class="hljs-number">1</span>/<span class="hljs-number">5</span>
<span class="hljs-number">17200</span>/<span class="hljs-number">17200</span> [==============================] - <span class="hljs-number">207</span>s <span class="hljs-number">12</span>ms/sample - loss: <span class="hljs-number">0.0096</span>
Epoch <span class="hljs-number">2</span>/<span class="hljs-number">5</span>
<span class="hljs-number">17200</span>/<span class="hljs-number">17200</span> [==============================] - <span class="hljs-number">202</span>s <span class="hljs-number">12</span>ms/sample - loss: <span class="hljs-number">0.0059</span>
Epoch <span class="hljs-number">3</span>/<span class="hljs-number">5</span>
<span class="hljs-number">17200</span>/<span class="hljs-number">17200</span> [==============================] - <span class="hljs-number">202</span>s <span class="hljs-number">12</span>ms/sample - loss: <span class="hljs-number">0.0058</span>
Epoch <span class="hljs-number">4</span>/<span class="hljs-number">5</span>
<span class="hljs-number">17200</span>/<span class="hljs-number">17200</span> [==============================] - <span class="hljs-number">200</span>s <span class="hljs-number">12</span>ms/sample - loss: <span class="hljs-number">0.0057</span>
Epoch <span class="hljs-number">5</span>/<span class="hljs-number">5</span>
<span class="hljs-number">17200</span>/<span class="hljs-number">17200</span> [==============================] - <span class="hljs-number">200</span>s <span class="hljs-number">12</span>ms/sample - loss: <span class="hljs-number">0.0057</span>
Test RMSE: <span class="hljs-number">0.0926447226146452</span>
</code></pre>
    <p class="normal">As in the previous section, a GA optimized a section of an ANN. You can use a GA for other components of an ANN. The sky is the limit! In fact, there is no limit. The potential of hybrid neural networks using GA or other algorithms to optimize their architecture or process takes your projects to another level!</p>
    <h1 id="_idParaDest-362" class="title">Summary</h1>
    <p class="normal">Evolutionary algorithms bring new light to AI's optimizing potential. In this chapter, we studied how heredity deeply affects population distribution. The impact of our environment can be measured through genetic mutations.</p>
    <p class="normal">Drilling down further, we focused on a class of GAs implementing a simulation of genetic transformations through many generations. We explored how a parent will transmit some genes, but how the selection of diverse genes from the generation population of genes will produce variations. A chromosome will inherit some genes but not others.</p>
    <p class="normal">The pressure of nature and our environment will take over. A fitness function evaluates a string of genes. Only the fittest will survive. The fittest genetic material will produce a crossover and mutation of the child, making it fitter for its environment.</p>
    <p class="normal">GAs can be used to represent strings of any type of data and also features of that data. The optimizing process can be applied to warehouses, transportation, and neural networks.</p>
    <p class="normal">Hybrid networks will no doubt expand in the years to come, taking DL to the next level. Using a genetic algorithm to optimize the architecture of an RNN paves the way to optimize the architecture of any DL, ML or AutoML architecture. For example, a hybrid neural network can use a genetic algorithm to optimize inputs with feature reduction or as the weight optimizing function of the network.</p>
    <p class="normal">Nature has provided us with invaluable tools to apply to our artificial network models. This chapter dealt with the invisible building blocks inside us. In the next chapter, <em class="italics">Neuromorphic Computing</em>, we will explore other components that enable us to adapt to our environment: neurons. We will explore how neural networks using biological models can solve complex problems.</p>
    <h1 id="_idParaDest-363" class="title">Questions</h1>
    <ol>
      <li class="list">A cell contains 42 chromosomes. (Yes | No)</li>
      <li class="list">A genetic algorithm is deterministic, not random. (Yes | No)</li>
      <li class="list">An evolutionary algorithm means that the program code evolves. (Yes | No)</li>
      <li class="list">It is best for a child to have the same genes as one of the parents even after many generations. (Yes | No)</li>
      <li class="list">Diversity makes the gene sets weaker. (Yes | No)</li>
      <li class="list">Building a neural network only takes a few lines, and the architecture always works. (Yes | No)</li>
      <li class="list">Building a neural network with a genetic algorithm can help optimize the architecture of the layers. (Yes | No)</li>
      <li class="list">Hybrid neural networks are useless since deep learning will constantly progress. (Yes | No)</li>
      <li class="list">Would you trust a genetic algorithm to make decisions for you? (Yes | No)</li>
      <li class="list">Would you trust a hybrid neural network to optimize the architecture of your network? (Yes | No)</li>
    </ol>
    <h1 id="_idParaDest-364" class="title">Further reading</h1>
    <ul>
      <li class="list"><a href="https://github.com/DEAP/deap"><span class="url">https://github.com/DEAP/deap</span></a></li>
      <li class="list"><a href="https://pypi.org/project/bitstring/"><span class="url">https://pypi.org/project/bitstring/</span></a></li>
    </ul>
  </div>
</body></html>